spyder/widgets/sourcecode/base.py

```diff
@@ -886,61 +886,66 @@ def __move_line_or_selection(self, after_current_line=True):
             cursor.movePosition(QTextCursor.NextBlock)
-            end_pos = cursor.position()
-            # check if on last line
-            if end_pos == start_pos:
-                cursor.movePosition(QTextCursor.End)
-                end_pos = cursor.position()
-                if start_pos == end_pos:
-                    cursor.endEditBlock()
-                    return
-                add_linesep = True
+        end_pos = cursor.position()
+        
+        # check if selection ends on the last line of the document
+        if cursor.atEnd():
+            if not cursor.atBlockStart() or end_pos == start_pos:
+                last_line = True
+                
+        # ------ Stop if at document boundary
+        
         cursor.setPosition(start_pos)
+        if cursor.atStart() and not after_current_line:
+            # Stop if selection is already at top of the file while moving up
+            cursor.endEditBlock()
+            self.setTextCursor(cursor)
+            self.__restore_selection(start_pos, end_pos)
+            return
+                
         cursor.setPosition(end_pos, QTextCursor.KeepAnchor)
-
+        if last_line and after_current_line:
+            # Stop if selection is already at end of the file while moving down
+            cursor.endEditBlock()
+            self.setTextCursor(cursor)
+            self.__restore_selection(start_pos, end_pos)
+            return
+        
+        # ------ Move text
+        
         sel_text = to_text_string(cursor.selectedText())
-        if add_linesep:
-            sel_text += os.linesep
         cursor.removeSelectedText()
-
-        if after_current_line:
-            text = to_text_string(cursor.block().text())
-            if len(text) == 0:
-                #If the next line is blank
-                sel_text = sel_text[0:-1]
-                sel_text = os.linesep + sel_text
-            if not text:
-                cursor.insertText(sel_text)
-                cursor.endEditBlock()
-                return
+        
+        if after_current_line:  # shift selection down
+            text = to_text_string(cursor.block().text())  
+            sel_text = os.linesep + sel_text[0:-1]  # move linesep at the start
+            cursor.movePosition(QTextCursor.EndOfBlock)
             start_pos += len(text)+1
-            end_pos += len(text)+1
-            cursor.movePosition(QTextCursor.NextBlock)
-            if cursor.position() < start_pos:
-                cursor.movePosition(QTextCursor.End)
-                sel_text = os.linesep + sel_text
-                end_pos -= 1
-        else:
-            cursor.movePosition(QTextCursor.PreviousBlock)
+            end_pos += len(text)
+            if not cursor.atEnd():
+                end_pos += 1
+        else:  # shift selection up
+            if last_line:
+                cursor.deletePreviousChar()  # remove last linesep...
+                sel_text = sel_text + os.linesep  # ...and add it to sel_text

```

Comment:
Please make a comment below the condition with something like `# Remove the last linesep and add it to the selected text` instead of the comments per line.